packages/json-rpc/ws-provider/src/ws-provider.ts

```diff
@@ -1,56 +1,185 @@
 import type { JsonRpcProvider } from "@polkadot-api/json-rpc-provider"
 import { getSyncProvider } from "@polkadot-api/json-rpc-provider-proxy"
 
-export const getInternalWsProvider =
-  (WebsocketClass: typeof WebSocket) =>
-  (uri: string, protocols?: string | string[]): JsonRpcProvider =>
-    getSyncProvider(async () => {
-      const socket = new WebsocketClass(uri, protocols)
+export enum WsEvent {
+  CONNECTING,
+  CONNECTED,
+  ERROR,
+  CLOSE,
+}
+export type WsConnecting = {
+  type: WsEvent.CONNECTING
+  uri: string
+  protocols?: string | string[]
+}
+export type WsConnected = {
+  type: WsEvent.CONNECTED
+  uri: string
+  protocols?: string | string[]
+}
+export type WsError = {
+  type: WsEvent.ERROR
+  event: any
+}
+export type WsClose = {
+  type: WsEvent.CLOSE
+  event: any
+}
+export type StatusChange = WsConnecting | WsConnected | WsError | WsClose
+export type WsJsonRpcProvider = JsonRpcProvider & {
+  switch: (uri?: string, protocol?: string[]) => void
+  getStatus: () => StatusChange
+}
+
+export interface GetWsProviderInput {
+  (
+    uri: string,
+    protocols?: string | string[],
+    onStatusChanged?: (status: StatusChange) => void,
+  ): WsJsonRpcProvider
+  (
+    uri: string,
+    onStatusChanged?: (status: StatusChange) => void,
+  ): WsJsonRpcProvider
+  (
+    endpoints: Array<string | { uri: string; protocol: string[] }>,
+    onStatusChanged?: (status: StatusChange) => void,
+  ): WsJsonRpcProvider
+}
+
+export const getInternalWsProvider = (
+  WebsocketClass: typeof WebSocket,
+): GetWsProviderInput => {
+  return (...args): WsJsonRpcProvider => {
+    let endpoints: Array<[string, string | string[]] | [string]> = []
+    let onStatusChanged: (status: StatusChange) => void = () => {}
 
+    if (typeof args[1] === "function") onStatusChanged = args[1]
+    if (Array.isArray(args[0]))
+      endpoints = args[0].map((x) =>
+        typeof x === "string" ? [x] : [x.uri, x.protocol],
+      )
+    else {
+      endpoints = [[args[0]]]
+      if (args[1]) {
+        if (args[1] !== onStatusChanged) endpoints[0][1] = args[1] as any
+        else if (args[2]) onStatusChanged = args[2] as any
+      }
+    }
```

Comment:
I think these two overloads will fail:

```ts
getWsProvider("uri", undefined, onStatusChanged);
getWsProvider("uri", "protocols", onStatusChanged);
```

The first one, arguably they could just use `getWsProvider('uri', onStatusChanged)`, but the types still allow to use that version.

For the second one, if I'm reading this correctly, it will go down through this else branch and...

- endpoints is assigned to `[["uri"]]`
- enters the `if (args[1])`
- enters the `if (args[1] !== onStatusChanged)`
- doesn't enter the next one, since it's behind an else.

Reply:
good catch! fixed! thanks!